File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-6.js
class Solution {
    /**
     * @param {string} s
     * @return {number}
     */
    lengthOfLongestSubstring(s) {
        let seen = new Set();
        let left = 0;
        let right = 0;
        let longest = 0;

        while (right < s.length) {
            while (seen.has(s[right])) {
                seen.delete(s[left]);
                left++;
            }
            seen.add(s[right]);
            longest = Math.max(longest, right - left + 1);
            right++;
        }
        return longest;
    }
}

File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-6.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        freq = {}
        maxf = 0
        left, right = 0, 0
        res = 0

        while right < len(s):
            freq[s[right]] = freq.get(s[right], 0) + 1
            maxf = max(freq.get(s[right]), maxf)

            while right - left - maxf >= k:
                freq[s[left]] -= 1
                left += 1

            
            res = max(res, right - left + 1)

            right += 1
        
        return res
